python/golem_task_api/testutils/dockertaskapiservice.py

```diff
@@ -0,0 +1,129 @@
+    """ Check the connection with Docker daemon """
+    try:
+        client = docker.from_env()
+        client.ping()
+    except Exception:
+        return False
+    return True
+
+
+def run_docker_container(
+        image: str,
+        work_dir: Path,
+        command: str,
+        port: int
+) -> docker.models.containers.Container:
+    """" Start a detached container """
+    if sys.platform == 'linux':
+        ports = {}
+    else:
+        ports = {port: port}
+
+    client = docker.from_env()
+    return client.containers.run(
+        image=image,
+        command=command,
+        volumes={
+            str(work_dir): {
+                'bind': f'/{constants.WORK_DIR}',
+                'mode': 'rw',
+            }
+        },
+        user=os.getuid(),
+        ports=ports,
+        detach=True,
+    )
+
+
+def get_docker_container_port_mapping(
+        container_id: str,
+        port: int,
+        vm_name: str,
+) -> Tuple[str, int]:
+    """ Retrieve running container's IP address and port """
+
+    client = docker.APIClient()
+    container_config = client.inspect_container(container_id)
+    net_config = container_config['NetworkSettings']
+
+    if sys.platform == 'darwin':
+        ip_address = '127.0.0.1'
+    elif sys.platform == 'win32':
+        ip_address = subprocess.check_output(['docker-machine', 'ip', vm_name])
+        ip_address = ip_address.decode('utf-8').strip()
+    else:
+        ip_address = net_config['Networks']['bridge']['IPAddress']
+
+    if sys.platform != 'linux':
+        port = int(net_config['Ports'][f'{port}/tcp'][0]['HostPort'])
+
+    if not ip_address:
+        raise RuntimeError(f'Unable to read the IP address of {container_id}')
+
+    return ip_address, port
+
+
+class DockerTaskApiService(TaskApiService):
+
+    def __init__(
+            self,
+            image: str,
+            work_dir: Path,
+            vm_name: Optional[str] = 'default'
+    ) -> None:
+        self._image = image
+        self._work_dir = work_dir
+        self._vm_name = vm_name
+        self._container = None
+
+    async def start(
+            self,
+            command: str,
+            port: int
+    ) -> Tuple[str, int]:
+        self._container = run_docker_container(
+            self._image,
+            self._work_dir,
+            command,
+            port)
+
+        return get_docker_container_port_mapping(
+            self._container.id,
+            port,
+            self._vm_name)
+
+    async def stop(self) -> None:
+        if not self.running():
+            return
+        self._container.stop()
+
+    def running(self) -> bool:
+        try:
+            self._container.reload()
+        except (AttributeError, docker.errors.APIError):
+            return False
+        return self._container.status not in ['exited', 'error']
+
+    async def wait_until_shutdown_complete(self) -> None:
+        print(f'Shutting down container with status: {self._container.status}')
```

Comment:
afair .status is a cached property, to read the most recent values you need to call another FN, can not find it so quickly but will look more

Reply:
It's `reload`. It is already being called in `running()`.